libopae++/include/opaec++/dma_buffer.h

```diff
@@ -0,0 +1,248 @@
+// POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+#include <memory>
+#include <cstdint>
+#include <vector>
+#include <initializer_list>
+#include <chrono>
+#include <thread>
+
+#include <opae/buffer.h>
+#include <opaec++/handle.h>
+
+namespace opae
+{
+namespace fpga
+{
+namespace types
+{
+
+class dma_buffer : public std::enable_shared_from_this<dma_buffer>
+{
+public:
+    typedef std::size_t size_t;
+    typedef std::shared_ptr<dma_buffer> ptr_t;
+
+    /** dma_buffer destructor.
+     */
+    virtual ~dma_buffer();
+
+    /** dma_buffer factory method - allocate a dma_buffer.
+     * @param[in] handle The handle used to allocate the buffer.
+     * @param[in] len    The length in bytes of the requested buffer.
+     * @return A valid dma_buffer smart pointer on success, or an
+     * empty smart pointer on failure.
+     */
+    static dma_buffer::ptr_t allocate(handle::ptr_t handle, size_t len);
+
+    /** Retrieve the virtual address of the buffer base.
+     */
+    volatile uint8_t * get() const { return virt_; }
+
+    /** Retrieve the handle smart pointer associated with
+     * this buffer.
+     */
+    handle::ptr_t get_owner() const { return handle_; }
+
+    /** Retrieve the length of the buffer in bytes.
+     */
+    size_t get_size() const { return len_; }
+
+    /** Retrieve the address of the buffer suitable for
+     * programming into the accelerator device.
+     */
+    uint64_t get_iova() const { return iova_; }
+
+    /** Divide a buffer into a series of smaller buffers.
+     *
+     * For each item sz found in sizes, create a new dma_buffer
+     * smart pointer whose size is sz and append the new smart
+     * pointer to the end of the returned vector. The sub-buffers
+     * are created in increasing order from the beginning of
+     * this dma_buffer. The parent buffer (this) of each sub-buffer
+     * is tracked so that it cannot be freed prior to a sub-buffer
+     * free.
+     *
+     * @param[in] sizes An initializer list of sizes for the sub-buffers.
+     * @return A std::vector of the sub-buffer pointers.
+     */
+    template <typename T>
+    std::vector<ptr_t> split(std::initializer_list<T> sizes)
+    {
+        std::vector<ptr_t> v;
+        size_t offset = 0;
+
+        v.reserve(sizes.size());
+
+        for (const auto &sz : sizes) {
+            ptr_t p;
+            p.reset(new dma_buffer(handle_,
+                                   sz,
+                                   virt_ + offset,
+                                   wsid_,
+                                   iova_ + offset,
+                                   shared_from_this()));
+            v.push_back(p);
+            offset += sz;
+        }
+
+        return v;
+    }
+
+    /** Write c to each byte location in the buffer.
+     */
+    void fill(int c);
+
+    /** Compare this dma_buffer (the first len bytes)
+     * to that held in other, using memcmp().
+     */
+    int compare(ptr_t other, size_t len) const;
```

Comment:
Should we overload this function with a version that doesn't include `len` parameter. It would compare:
* both sizes of both buffers
* both buffers for the entire length

Reply:
In the spirit of keeping the API simple, why don't we save that for a higher-level class.